### ISSUE TITLE
Occasional incorrect error: Duplicate base class not allowed

### ISSUE THREAD

**Issue by RomainMuller** [microsoft/pyright#3781]
**Describe the bug**

Running `pyright` on a generated codebase, I am occasionally getting the `error: Duplicate base class not allowed` error, despite the class in question has two **distinct** bases.

The particularity of the class that triggers this error is that it has two parents that have the same name in their respective modules, but originate from different modules... schematically:

```py
from ..submodule import Bar as _Bar_aliased

class Bar:
    ...

class Foo(Bar, _Bar_aliased): # E: Duplicate base class not allowed
    ...
```

Interestingly, this error seems to either consistently happen, or consistently NOT happen in a given environment, and I have so far not been able to identify what differentiates these environments.

I have a GitHub Workflow test matrix where the `pyright` execution succeeded (no errors) in all `ubuntu-latest` runs, but failed in the `macos-latest` and `windows-latest` environments. I've definitely had both succeeding and failing runs on my machine (which happens to be a Mac).

**To Reproduce**

I have tried to craft a minimal reproduction for this issue but have so far not succeeded. As mentioned, I've seen the checker succeed and fail on the exact same code, exact same machine...

I've also tried to set up a Docker image that would consistently demonstrate the issue, to no avail.

So I've pushed the code with which I'm having problems to a Github repository, in hopes this can help someone to investigate: https://github.com/RomainMuller/pyright-dublicate-base-error. The unexpected error is showing up on https://github.com/RomainMuller/pyright-dublicate-base-error/blob/9f0067283b44fae09b5b3a513efa12b411513135/generated-code/jsii-calc/python/src/jsii_calc/module2692/submodule2/__init__.py#L58.

**Expected behavior**

Expected this code to not produce a `Duplicate base class not allowed` since there is no duplicate base class here.

**Screenshots or Code**
<img width="947" alt="Capture d’écran 2022-08-03 à 16 45 37" src="https://user-images.githubusercontent.com/411689/182637893-7524cc7d-909e-46f1-925f-ca9ad26e8ac2.png">

**VS Code extension or command-line**
I am running `pyright` from the command-line (installed via `npm`, version `1.1.265`).
I am also observing the same via the Pylance VSCode extension (not consistently as well).

**Comment by RomainMuller** [comment]
The PR on which I am trying to introduce the `pyright` test (https://github.com/aws/jsii/pull/3694) currently has validations failing on this particular issue on both `macos` and `windows`, but not on any other flavor: https://github.com/aws/jsii/runs/7653167638

**Comment by RomainMuller** [comment]
Note w/r/t the repro repository I've provided, I just tried it and:
- I do _not_ get the error when running `pyright` from the CLI
- I **do** get the error in VSCode when opening the offending file w/ Pylance enabled

**Comment by RomainMuller** [comment]
(Note that ignoring `reportGeneralTypeIssues` here is not an option, as the reason I am running `pyright` here in the first place is to detect issues where the code-generator produces incorrect or otherwise invalid typings... and `reportGeneralTypeIssues` - like most diagnostics categories in `pyright` is too broad)

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem, and I'm looking into it now.

**Comment by erictraut** [comment]
I've found the cause of the problem I was able to repro, but I don't think it explains all of the symptoms you have observed.

Here's what's going on. The reason you're seeing a "duplicate base class not allowed" error is because the type checker thinks that `Bar` and `_Bar_ec7eccad` are the same class. They are declared with the same name (`Bar`), but they are defined in different files. They should have different "fully-qualified names", and the type evaluator should be able to tell them apart based on this difference. The fully-qualified names are based on the module path for the source file that defines the class. For example, if the file is `x/y/z/__init__.py`, then the module path is `x.y.z`, and the fully-qualified name of the type `Bar` is `x.y.z.Bar`. The module path is computed based on import resolution paths because it is relative to the import path. For example, if your import resolution paths consist of the list `['.', './x/y']`, then the file `./x/y/z/__init__.py` will have a module path of `x.y.z`. If you provide the import resolution paths in the opposite order (`['./x/y', '.']`), the same source file will have a module path of `z`. So the current import resolution paths (and their order!) can change the module path for a source file.

There are situations where pyright cannot compute a module path for a source file. For example, if your import resolution path is `'.'` and the file path is `./a-b/c/__init__.py`, the file path segment `a-b` cannot be represented in a module path because `-` is an illegal character. Pyright needs to handle this case because you can open Python files from any location on disk and view them using VS Code. In cases where a valid module path cannot be computed, pyright internally sets the module path to an empty string. That's what was happening on your case with the module path for both `submodule1/__init__.py` and `submodule2/__init__.py`. Both of these submodules fall under a path that includes the path segments `generated-code` and `jsii-calc`, which cannot be represented in a module path. Consequently, the two `Bar` classes defined in these respective files both have fully-qualified names of `.Bar`, and the type checker couldn't distinguish between them.

Once you select the new venv (the one created in your `prepare-venv.sh` script) and complete the editable installs, the import resolution paths for the environment are updated to include the directory `./generated-code/jsii-calc/python/src/jsii_calc`, and the module path for `submodule1` and `submodule2` can now be computed relative to these new import paths.

The import resolution paths used by pyright include: the root directory for the project (`'.'`), plus any additional paths in `sys.path` for the currently-selected Python interpreter, plus any additional paths specified in the "extraPaths" setting. You don't have any "extraPaths" specified in your pyproject.toml file, so that doesn't apply in your case. So the paths that we're interested in here are those that come from `sys.path` for your currently-selected Python interpreter. When you run the CLI version of pyright, it assumes that you've already specified the desired python interpreter before executing pyright, and it simply shell execs `python3` and runs a short script that retrieves the value of `sys.path`. When running pyright (or pylance) in VS Code, the python interpreter depends on what you have selected in the editor (specifically, in the popup menu provided by the Python extension to VS Code).

If you change the selected Python interpreter in VS Code, the import resolution paths will change, and this can in turn change the module paths for source files that pyright had previously enumerated.

The bug that I've fixed can be seen here with [this commit](https://github.com/microsoft/pyright/commit/cdba65d4c8931a2a2b2dfa26df0f483973f1d3d2). The problem was that pyright was not recomputing the module path based on the updated import resolution paths. So even if you selected a new Python interpreter (and hence updated the import resolution paths), it was retaining the old module path. The fix is to update the module path and not rely on the old cached value.

I was previously able to repro the problem you're seeing by selecting some other Python interpreter (not the one in the local venv) and then switching to the local venv. The "duplicate base class" error should go away after switching to the local venv, but the errors remained because of the stale cache. With my bug fix in place, the error goes away after switching to the local venv.

With this fix, I'm no longer able to repro the problem. However, I can't yet explain why you would see the problem sporadically when running the CLI version. Let me know if you have any thoughts on this.

**Comment by RomainMuller** [comment]
I don't know how `pyright` internally is designed, but the sporadic nature of the issue made me think there was perhaps a race condition, and the end behavior may have depended on the order in which files were processed...

I was initially inclined to run `pyright` separately on each one of the generated packages, which would have likely prevented the presence of the `-` segments in the workspace-relative file path, however since it is not possible to specify `--venv` via CLI argument, this would have required me to modify the `pyproject.toml` files in there which I'd rather not do...

---

In terms of other "peculiar" things my test harness is doing, which _might_ contribute to triggering edge-case behaviors... the code is normally being generated in a temporary directory (and it's quite probable the structure of the temporary directory path is significantly different between platforms -- on Linux it'll be something like `/tmp/jsii-pacmak-pyright-<random>`, on MacOS it'll be something like `/var/folders/nx/81rqfpmd20d_g7j3y4bztv800000gs/T/jsii-pacmak-pyright-<random>`, and God know what it's like on Windows)... Maybe the long MacOS path also gets into an edge case?

Another aspect (possibly in favor of the race condition / order of processing argument) is that each of these platforms has a different file system, and the order in which directory listings are returned may be different... Which may cause files to be processed in different order in different platforms, and the caching you've fixed might actually also have bit in this case?

**Comment by erictraut** [comment]
My first thought when I saw your bug report was that this might be related to the fact that file systems are case-sensitive in Linux and case-insensitive in MacOS and Windows. But that doesn't explain the symptoms you're seeing.

I also thought about directory listing order. If the behavior is as nondeterministic as you say, that's a good theory.

Pyright's core logic is synchronous and single-threaded, so this isn't related to threading.

Long paths shouldn't matter. Paths with characters that cannot be represented in a module path (such as `-` or `@`) will affect the behavior, but it looks like you're using `-` in paths on all three platforms. So that doesn't explain the difference.

I think the best course of action here is to release the next version of pyright with my fix and see if you are able to repro the problem. If not, then I'll assume this was fully explained by the bug I fixed. If you're still able to repro the issue, I'll investigate further.

**Comment by RomainMuller** [comment]
>I think the best course of action here is to release the next version of pyright with my fix and see if you are able to repro the problem. If not, then I'll assume this was fully explained by the bug I fixed. If you're still able to repro the issue, I'll investigate further.

100% agree here!

**Comment by erictraut** [comment]
I just released pyright 1.1.266. Please give it a try and see if you can still repro the problem.

**Comment by RomainMuller** [comment]
Unfortunately, with `1.1.266` I still get the same issue on Windows and MacOS on my PR validation... 😔

```
jsii-pacmak:     #STDOUT> C:\Users\RUNNER~1\AppData\Local\Temp\jsii-pacmak-pyright-T2Qrjh\jsii-calc\python\src\jsii_calc\module2692\submodule2\__init__.py
jsii-pacmak:     #STDOUT>   C:\Users\RUNNER~1\AppData\Local\Temp\jsii-pacmak-pyright-T2Qrjh\jsii-calc\python\src\jsii_calc\module2692\submodule2\__init__.py:58:16 - error: Duplicate base class not allowed (reportGeneralTypeIssues)
jsii-pacmak:     #STDOUT> 1 error, 0 warnings, 0 informations·
jsii-pacmak:     #STDOUT> Completed in 4.552sec
```

**Comment by RomainMuller** [comment]
I can confirm that renaming the sub-directories in my workspace that include `@` or `-` to versions with only `.` and `_` appears to make the issue go away. Still not clear to me why some platforms/tests appear to not fall into the trap...?

**Comment by erictraut** [comment]
I'm not able to repro the problem that you're seeing with your exact repro steps, but I was able to find another way to reliably repro it. I've made pyright's code more resilient so it can handle cases where there are multiple modules within your project's root that are under subdirectories with names that contain invalid characters for an import statement (e.g. "-"). So this should resolve the problem once and for all.

This change will be in the next release of pyright.

**Comment by erictraut** [comment]
This is included in pyright 1.1.270, which I just published. It will also be included in a future release of pylance.
